text/0328-polyglot-assert.md

```diff
@@ -0,0 +1,336 @@
+---
+rfc pr: https://github.com/aws/aws-cdk-rfcs/pull/330
+tracking issue: https://github.com/aws/aws-cdk-rfcs/issues/328
+---
+
+# Polyglot Assert
+
+The `@aws-cdk/assert` module is only available to javascript users.
+
+The new polyglot assert module - `@aws-cdk/assertions` - provides the same set of
+functionalities to users in all CDK supported languages.
+
+## Working Backwards
+
+### CHANGELOG
+
+**feat:** announcing assertions: the assert library is now available in all CDK supported languages.
+
+### README
+
+<!--BEGIN STABILITY BANNER-->
+
+---
+
+![cdk-constructs: Stable](https://img.shields.io/badge/cdk--constructs-stable-success.svg?style=for-the-badge)
+
+---
+
+<!--END STABILITY BANNER-->
+
+> NOTE: This module contains *beta APIs*.
+>
+> All beta symbols are suffixed with the `Beta<n>`. When we have backwards
+> incompatible change, we will create a new symbol with a `Beta<n+1>` suffix
+> and deprecate the `Beta<n>` symbol.
+---
+This module allows asserting the contents of CloudFormation templates.
+
+To run assertions based on a CDK `Stack`, start off with -
+
+```ts
+const stack = new cdk.Stack(...)
+...
+const assert = TemplateAssertionsBeta1.fromStack(stack);
+```
+
+Alternatively, assertions can be run on an existing CloudFormation template -
+
+```ts
+const template = fs.readFileSync('/path/to/template/file');
```

Comment:
> I have to say, I don't understand why would anyone use this. What's the use cases for this?

use case: Run any of our assertions on a CloudFormation template not generated by the CDK. 
I don't see why this should be used only on template generated from a CDK `Stack`. There's literally no additional effort in supporting this.

> Should we not have a way to create an Assertions object from the Cloud Assembly that's returned from `App.synth()`?

Reading between the lines, it seems you're assuming the use case is to pass a CDK synthesized template generated in the cloud assembly. It's not.

Reply:
> > Should we not have a way to create an Assertions object from the Cloud Assembly that's returned from `App.synth()`?
> 
> Reading between the lines, it seems you're assuming the use case is to pass a CDK synthesized template generated in the cloud assembly. It's not.

Not sure what you mean exactly. Can we please add an API based on the Cloud Assembly? We need it ourselves, as we use it today ([one of many examples](https://github.com/aws/aws-cdk/blob/093ebcb942cd789b2516b4b3f577db3c9c1766c4/packages/%40aws-cdk/core/test/cfn-resource.test.ts#L63-L72)).

Reply:
In the example provided, why do we do this instead of simply `assertMatchTemplate(stack, { ... })`?

Reply:
Oh I get it now. Our current assert library depends on core and hence cannot be used in the unit tests for core.
The same restrictions will apply to the new assert library too.

Do you have an example that's not in core?

Reply:
> > I have to say, I don't understand why would anyone use this. What's the use cases for this?
> 
> use case: Run any of our assertions on a CloudFormation template not generated by the CDK.
> I don't see why this should be used only on template generated from a CDK `Stack`. There's literally no additional effort in supporting this.

I think that, if you want to support that, the input should be the path to the file, not the contents of the file. Something like this:

```ts
const assert = TemplateAssertionsBeta1.fromFile('/path/to/template/file');
```

The reason for this is that parsing CloudFormation-flavored YAML is very non-trivial ([see here](https://github.com/aws/aws-cdk/blob/252dfa2f84f24ef57ab632e8ee5092544c850a5f/packages/%40aws-cdk/yaml-cfn/lib/yaml.ts#L21-L30)), and you don't want your customers to have to reinvent that.

Reply:
> Oh I get it now. Our current assert library depends on core and hence cannot be used in the unit tests for core.
> The same restrictions will apply to the new assert library too.
> 
> Do you have an example that's not in core?

Sure. [Here's one](https://github.com/aws/aws-cdk/blob/252dfa2f84f24ef57ab632e8ee5092544c850a5f/packages/%40aws-cdk/aws-s3-assets/test/asset.test.ts#L50-L72).

We have many, many tests like these.

Reply:
Sure, we can add it as and when we need it, just like any other library. 
The plan is to migrate (some) modules to start using these assertion library and eventually all modules will use it. This will force us to build this up iteratively.

That's not the focus of this RFC.